Implement endpoint to download dir contents as zip

rpschoenburg opened the pull request:
Implements new client and server modules to support downloading dir contents as a zip file, rather than just as compressed tarballs. This requires the previous batch_download logic to be split into two separate functions, one for tarballs and the other for zip files, and the use of the async_zip library.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Download repository directories as ZIP streams (1 GB max).
  * Import files from external URLs into repositories.
  * Upload ZIP archives to import content into repositories.
  * Client/CLI: added directory-as-ZIP download and ZIP upload/import helpers.

* **Chores**
  * Reorganized server routes to expose dedicated import/export endpoints.
  * Server streaming and ZIP handling implemented; tests updated.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rpschoenburg committed:
Implement endpoint to download dir contents as zip

Implements new client and server modules to support downloading dir
contents as a zip file, rather than just as compressed tarballs. This
requires the previous batch_download logic to be split into two separate
functions, one for tarballs and the other for zip files, and the use of
the async_zip library.

rpschoenburg committed:
Correct url for api::client::dir::list

rpschoenburg committed:
CodeRabbit suggestion

rpschoenburg committed:
remove comment

rpschoenburg committed:
Upload zip download to recreate dir structure

rpschoenburg committed:
Add unit test for dir zip download

rpschoenburg commented on oxen-rust/src/lib/src/api/client/dir.rs:
```diff
@@ -30,7 +32,7 @@ pub async fn list(
     let revision = revision.as_ref();
     let path = path.as_ref().to_string_lossy();
-    let uri = format!("/dir/{revision}/{path}?page={page}&page_size={page_size}");
+    let uri = format!("/dir/download/{revision}/{path}?page={page}&page_size={page_size}");
```
Not sure about this one. I made the change so download_dir_as_zip wouldn't match into the old endpoint. Any better ideas for a name?

rpschoenburg commented on oxen-rust/src/server/src/services/dir.rs:
```diff
@@ -4,5 +4,13 @@ use actix_web::Scope;
+        .route(
+            "/download_zip/{resource:.*}",
+            web::get().to(controllers::dir::download_dir_as_zip),
+        )
```
This is the same issue as above

rpschoenburg committed:
CodeRabbit Suggestions

rpschoenburg commented on oxen-rust/src/lib/src/api/client/dir.rs:
```diff
@@ -93,6 +95,44 @@ pub async fn get_dir(
+            StreamReader::new(stream.map(|result| result.map_err(std::io::Error::other)));
+
+        let mut file = File::create(local_path).await?;
+        let size = tokio::io::copy(&mut reader, &mut file).await?;
```
changed in latest commit

rpschoenburg commented on oxen-rust/src/lib/src/api/client/dir.rs:
```diff
@@ -1,14 +1,16 @@
+use async_std::prelude::StreamExt;
```
changed in latest commit

rpschoenburg replied:
changed in latest commit

rpschoenburg committed:
Resolve dependency conflict

jcelliott commented on oxen-rust/Cargo.lock:
```diff
@@ -3184,7 +3189,7 @@ dependencies = [
  "option-ext",
  "redox_users 0.5.2",
- "windows-sys 0.61.2",
+ "windows-sys 0.59.0",
```
@rpschoenburg any idea why several of these versions were downgraded here?

jcelliott commented on oxen-rust/src/server/src/services/dir.rs:
```diff
@@ -4,5 +4,13 @@ use actix_web::Scope;
 
 pub fn dir() -> Scope {
-    web::scope("/dir").route("/{resource:.*}", web::get().to(controllers::dir::get))
+    web::scope("/dir")
```
@rpschoenburg I know we talked about adding the download as zip endpoint under `/dir`, but after looking at this more I don't think it makes sense here. For one thing, this wouldn't be backwards compatible, but semantically I think it should live in its own namespace as well. We've also overloaded the `/file` namespace with "import"-type actions. What do you think about making an `/import` and `/export` namespace? The download as zip function could go under `/export/{resource:.*}` and later we can move some of those other operations to an analogous `/import/{resource:.*}` endpoint.

jcelliott commented on oxen-rust/src/lib/src/constants.rs:
```diff
@@ -182,8 +182,9 @@ pub const MAX_RETRIES: usize = 5;
 /// Allow up to N concurrent upload tasks
 pub const MAX_CONCURRENT_UPLOADS: usize = 30;
-// Retry and back off of requests N times
-/// Retry and back off of requests N times
```
Not sure what happened here. Looks like this comment was for the const below originally. Let's also explicitly document the size in human-readable form in the comment so it's easy to scan. (e.g., "100MiB").

I also think it would make sense to make this limit higher. Model weights are likely going to be larger than 100MB. Let's start with 1GB.

Have you done any testing on different size exports? Want to make sure we're not going to hit performance issues with larger downloads. Eventually we might need to switch to async export jobs, but as long as we can handle reasonable size model weights we shouldn't need to do that yet.

rpschoenburg committed:
Revert downgraded dependencies

rpschoenburg committed:
Migrate import and export endpoints to new modules

Creates the new import and export service and controller modules and
migrates the upload and download zip logic to these modules. Does not
deprecate the old endpoints

rpschoenburg committed:
Merge branch 'main' into API-80/download_dir_as_zip

rpschoenburg committed:
Update endpoint names

rpschoenburg committed:
Linters

rpschoenburg committed:
Update uri

rpschoenburg commented on oxen-rust/src/lib/src/api/client/dir.rs:
```diff
@@ -93,6 +95,59 @@ pub async fn get_dir(
+        let err = format!("try_download_dir_as_zip failed to send request {url}");
+        Err(OxenError::basic_str(err))
+    }
+}
```
We could potentially change these, but it's pretty tangential. The if let Ok(res) pattern is used in a lot of places in our API modules

rpschoenburg commented on oxen-rust/src/lib/src/api/client/dir.rs:
```diff
@@ -93,6 +95,59 @@ pub async fn get_dir(
+    } else {
+        let err = format!("try_download_dir_as_zip failed to send request {url}");
+        Err(OxenError::basic_str(err))
+    }
```
See above suggestion

rpschoenburg committed:
Create new import and export API modules

rpschoenburg committed:
update python repo

rpschoenburg committed:
Fix utoipa annotations
